### ISSUE TITLE
Disconnected entities not recognized as existing

### ISSUE THREAD

**Issue by patrikrazem** [dotnet/efcore#13051]
I am attempting a simple scenario of adding a disconnected entity, retrieved from one DbContext, to a different entity that is saved within another DbContext.

I have a basic DbContext implementation:

```cs
public class EfTestDbContext : DbContext
{
    public DbSet<Book> Books { get; set; }
    public DbSet<Author> Authors { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder.UseInMemoryDatabase();
    }
}
```

and two simple entities:

```cs
public class Author
{
    public int ID { get; set; }
    public string Name { get; set; }
}

public class Book
{
    public int ID { get; set; }

    public int AuthorID { get; set; }
    public Author Author { get; set; }
    
    public string Name { get; set; }
}
```

and I am creating and saving the `Author` entity within one context, then adding it to the `Book` entity within the other context:

```cs
static void Main(string[] args)
{
    Author existingAuthor;
    using (var ctx = new EfTestDbContext())
    {
        existingAuthor = new Author() { Name = "First author" };

        ctx.Add(existingAuthor);
        ctx.SaveChanges();
        Console.WriteLine($"Saved author ({existingAuthor.Name})");
    }

    using (var ctx = new EfTestDbContext())
    {
        var book1 = new Book()
        {
            Name = "First book",
            Author = existingAuthor
        };

        Console.WriteLine($"Added author ({existingAuthor.Name}) to first book");
        Console.WriteLine($"Author is existing: {ctx.Entry(existingAuthor).IsKeySet}");

        ctx.Add(book1);
        ctx.SaveChanges();
    }
}
```

Although the contexts are separate and the Author entity is disconnected once the first context is disposed, the `existingAuthor` entity is recognized as existing, since `ctx.Entry(existingAuthor).IsKeySet` returns `True`.

I have read the document at https://docs.microsoft.com/en-us/ef/core/saving/disconnected-entities and from what I understand, the `IsKeySet` value should be enough to let EF know that the entity already exists in the database and it should not attempt to insert it.

In this example I'm using the In-Memory provider, but I've attempted this with SQLite and the result has been the same.

Is this a bug, or am I doing something wrong? Have I understood the document wrong and should I be manually and explicitly tracking which entities  have a `IsKeySet` value set and which do not?

EF Core version: 2.1.1.

**Comment by footysteve** [comment]
Hi, perhaps I am misunderstanding your question but this,

"Although the contexts are separate and the Author entity is disconnected once the first context is disposed, the existingAuthor entity is recognized as existing, since ctx.Entry(existingAuthor).IsKeySet returns True."

appears to contradict your question and in fact, Disconnected entities ARE recognized as existing.

Looking at the code, without testing it out, this is the expected result, no?

**Comment by skyflyer** [comment]
Hi!

I was under the impression that `IsKeySet` returns `true` only when EF's change tracker would "assume" that the entity is being updated instead of added. 

Perhaps we're using EF Core in a way it was not meant to be used.

----

We created a couple of easy-to-run tests, which are available here:
https://github.com/skyflyer/eftest

There are three branches, each with their own specific approach:

### Using object references - `master` branch

This approach is the most straightforward, which uses object references to assign related entities and this breaks in `Changing author of the book` section of the program, when:

1. an existing author, fetched from a different (now disposed) DbContext, is set to
1. an existing book is fetched from the repository
1. and the book is being saved by `SaveChanges`

<details>
<summary>Relevant snippet</summary>

```c#
            using (var ctx = new EfTestDbContext())
            {
                Console.WriteLine($"Changing author of the book");
                var bookToChange = ctx.Books.First();
                bookToChange.Author = existingAuthor;
                Console.WriteLine($"  Changed book: {bookToChange.ID} - {bookToChange.Name} written by {bookToChange.Author.Name}");
                Console.WriteLine($"  Existing author's `IsKeySet`: {ctx.Entry(existingAuthor).IsKeySet}");

                ctx.Entry(existingAuthor).State = EntityState.Unchanged;
                ctx.SaveChanges();
                Console.WriteLine($"  Saved changes");
            }
```

</details>

I believe this could be a bug in EF Core's change tracker?

### Using Entity `ID` fields for associations - `using-ids` branch

This approach assigns related entities through `book.AuthorId = author.Id` properties on the parent entity and this approach works, but is not as convenient, especially when you would need access to the full related entity for some sort of validation that might occur... This is an acceptable workaround for us at the moment.

<details>
<summary>Relevant snippet</summary>

```c#
            using (var ctx = new EfTestDbContext())
            {
                Console.WriteLine($"Changing author of the book");
                var bookToChange = ctx.Books.First();
                bookToChange.AuthorID = existingAuthor.ID;
                Console.WriteLine($"  Changed book: {bookToChange.ID} - {bookToChange.Name} written by {bookToChange.AuthorID}");

                ctx.SaveChanges();
                Console.WriteLine($"  Saved changes");
            }
```
</details>

### Using explicit EntityState modifiers - `explicit` branch 

This approach is a bit awkward, since you'd have to track entity states manually and is IMHO also quite error prone. I'm talking about `ctx.Entry(existingAuthor).State = EntityState.Unchanged;` and this approach works without errors.

<details>
<summary>Relevant snippet</summary>

```c#
            using (var ctx = new EfTestDbContext())
            {
                Console.WriteLine($"Changing author of the book");
                var bookToChange = ctx.Books.First();
                bookToChange.Author = existingAuthor;
                Console.WriteLine($"  Changed book: {bookToChange.ID} - {bookToChange.Name} written by {bookToChange.Author.Name}");
                Console.WriteLine($"  Existing author's `IsKeySet`: {ctx.Entry(existingAuthor).IsKeySet}");

                ctx.Entry(existingAuthor).State = EntityState.Unchanged;
                ctx.SaveChanges();
                Console.WriteLine($"  Saved changes");
            }
```

</details>

**Comment by skyflyer** [comment]
Relevant log (with sensitive logging enabled) for the failing scenarion:

```
Context 'Book' started tracking '{ID: 1}' entity with key 'EfTestDbContext'.
  Changed book: 1 - First book written by An author
DetectChanges starting for 'EfTestDbContext'.
Navigation property 'Book.Author' for entity with key '{ID: 1}' detected as changed.
The 'Book' entity with key '{ID: 1}' tracked by 'EfTestDbContext' changed from 'Unchanged' to 'Modified'.
Foreign key property 'Book.AuthorID' detected as changed from '2' to '1' for entity with key '{ID: 1}'.
Context 'Author' started tracking '{ID: 1}' entity with key 'EfTestDbContext'.
DetectChanges completed for 'EfTestDbContext'.
  Existing author's `IsKeySet`: True
SaveChanges starting for 'EfTestDbContext'.
DetectChanges starting for 'EfTestDbContext'.
DetectChanges completed for 'EfTestDbContext'.
An exception occurred in the database while saving changes for context type 'eftest.EfTestDbContext'.
System.ArgumentException: An item with the same key has already been added. Key: 1
```

**Comment by skyflyer** [comment]
I also performed tests with the latest version of EF (`release/2.2` branch) and the results are the same.

I'm out of my depth in examining the true reason, but I tracked it to `NavigationFixer.cs:191` where this existing entity is attached as `Added` when traversing navigation property of the parent (`Book`) object. I noticed that change tracker is searching for existing entities in the context in order to determine whether an entity should be added or not (which contradicts with what is written in https://docs.microsoft.com/en-us/ef/core/saving/disconnected-entities or more probably, I'm misreading it):

https://docs.microsoft.com/en-us/ef/core/saving/disconnected-entities#with-auto-generated-keys 
> ... On the other hand, if the key value has been set, then it must have already been previously saved and now needs updating. ...

That lead me to manually attach the disconnected entity using `ctx.Attach(existingAuthor)` and then everything worked -- this approach is possible in simple example such as the provided test, but quite hard to achieve with repository implementations in different contexts.

So the puzzle remains - is there something special we need to tell EF that it should detected the entities correctly as modified?

**Comment by ajcvickers** [comment]
@patrikrazem When using `Add`, all entities are put in the Added state regardless of key value. This enables explicit key values to be used for inserts even when the key value is setup to be store-generated. When using `Attach` or `Update` with store generated keys, if the key value is not set, then the entity is marked as Added because it can't possibly be Unchanged or Modified.

**Comment by skyflyer** [comment]
@ajcvickers, what you write makes perfect sense. But... we're talking about "implicitly" attaching existing entities through navigation properties, as can be seen in the example. How should one handle such a scenario?

Are the only two approaches:

1) using `entity.RelatedId = relatedDisconnectedEntity.Id` or
2) using `ctx.Attach(relatedDisconnectedEntity)` and then `entity.Related = relatedDisconnectedEntity`?

As I mentioned before, I was under the impression (docs) that EF would detect related entity as "modified", since it has a key set.

In other words, I expected the example [here](https://github.com/skyflyer/eftest/blob/master/Program.cs#L65) to work.

**Comment by ajcvickers** [comment]
@skyflyer I was basing my answer on the original post, with specifically this code:
```C#
static void Main(string[] args)
{
    Author existingAuthor;
    using (var ctx = new EfTestDbContext())
    {
        existingAuthor = new Author() { Name = "First author" };

        ctx.Add(existingAuthor);
        ctx.SaveChanges();
        Console.WriteLine($"Saved author ({existingAuthor.Name})");
    }

    using (var ctx = new EfTestDbContext())
    {
        var book1 = new Book()
        {
            Name = "First book",
            Author = existingAuthor
        };

        Console.WriteLine($"Added author ({existingAuthor.Name}) to first book");
        Console.WriteLine($"Author is existing: {ctx.Entry(existingAuthor).IsKeySet}");

        ctx.Add(book1);
        ctx.SaveChanges();
    }
}
```
This is "attaching" a graph using Add.

The case you point to is the behavior of DetectChanges, which is not what the original post was about. When a new entity is associated with an already tracked graph, and that change is then detected by EF, then the new entity will always be in the Added state.

**Comment by ajcvickers** [comment]
@skyflyer There are two scenarios here:
* Adding/Attaching/Updating a graph of entities by explicitly calling one of the Add/Attach/Update methods. In this case, you are giving EF an explicit command to tell it how to track these entities.
* Modifying a graph of entities that are being tracked by EF, in your example by changing a reference in the graph to point to a different entity than it did before. In this case, EF doesn't know that that change has been made until it scans the graph of tracked entities to find what has changed--this is done by the [DetectChanges method](https://blog.oneunicorn.com/2012/03/10/secrets-of-detectchanges-part-1-what-does-detectchanges-do/). Now if the "new" entity found in this process is not tracked, then it will be automatically put in the Added state. This is because a) we don't know just because it has a key value that it is existing--see above, and b) the most common case for DetectChanges is that the detected entity is new and should be inserted.

**Comment by skyflyer** [comment]
@ajcvickers, I have observed what you write in practice, so it makes perfect sense. The reason Patrik opened this ticket is because we were actually investigating why this does not work according to the docs. I suppose we could agree that the docs are a little misleading. Take this paragraph for example (emphasis mine):

> The value of an automatically generated key **can often be used to determine** whether an entity needs to be inserted or updated. If the key has not been set (that is, it still has the CLR default value of null, zero, etc.), then the entity must be new and needs inserting. On the other hand, **if the key value has been set, then it must have already been previously saved and now needs updating**.  *In other words, if the key has a value, then the entity was queried, sent to the client, and has now come back to be updated.*

Anyhow, now that we know what is going on behind the scenes, we're all set. I was afraid we were doing something wrong or stumbling on a bug, but it has obviously been designed that way.

I appreciate your time responding to this issue. Thanks!
